libcontainer/container_linux.go

```diff
@@ -1674,13 +1674,34 @@ func (c *linuxContainer) updateState(process parentProcess) (*State, error) {
 	return state, nil
 }
 
-func (c *linuxContainer) saveState(s *State) error {
-	f, err := os.Create(filepath.Join(c.root, stateFilename))
+func (c *linuxContainer) saveState(s *State) (retErr error) {
+	tmpFile, err := ioutil.TempFile(c.root, "state-")
 	if err != nil {
 		return err
 	}
-	defer f.Close()
-	return utils.WriteJSON(f, s)
+
+	defer func() {
+		if retErr != nil {
+			tmpFile.Close()
+			os.Remove(tmpFile.Name())
+		}
+	}()
+
+	err = utils.WriteJSON(tmpFile, s)
+	if err != nil {
+		return err
+	}
+	err = tmpFile.Sync()
```

Comment:
Do you need this?

Reply:
Yes, we want to sync the file to make sure it is written to disk.

Reply:
That applies if the root isn't a tmpfs.

Reply:
I can drop it if we think we don't want to support that, but I don't see big downsides in keeping it.

Reply:
Sync may block, and this file does not need persistency

Reply:
The file contents, even if not synced to the disk, will be read back from the page cache the way it was written. The only thing that fsync helps with is when the host will lose power (or the kernel will crash) before writing the file to disk. I am not sure this is needed.

Reply:
Dropped the sync.